### PR TITLE
Support SM2 certificate verification

### PR DESCRIPTION
**Pull request by InfoHunter**
PR #6837 stalled for a long time so I open this PR to add the support of SM2 certificate verification. The main different between this PR and #6837 is I added the ability of specifying the SM2 ID rather than hardcoding it into "1234567812345678".

Several changes made to X509 code to achieve this, including an extra field carrying the ID value was added to X509 structure and the apps/verify command was also changed to let user set the ID (although the user is limited to only specify an ASCII string).

The usage of the `verify` app is something as this:

`openssl verify -sm2-id 1234567812345678 -CAfile sm2-ca.crt sm2.crt`

This is WIP status, so doc and test case are not updated yet. Please review this.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by t-j-h** [comment]
The sm2_id is not restricted to ascii/text only data - it should be able to be handled as a hexstring input as well so that arbitrary values can be represented - not just limited to literal ascii strings.

Other than that comment the rest is looking good to me.

**Comment by InfoHunter** [comment]
Did you suggest to support binary input in `verify` application?

**Comment by InfoHunter** [comment]
Currently the API supports that, but not the command line utility.

**Comment by mlindner** [comment]
Assuming this makes it in, can this be backported into 1.1.1?

**Comment by t-j-h** [comment]
Yes I am suggesting you allow input not limited to straight ascii in the command line.
It could be handled as a hex string usage.
Just like the underlying capability - this is not specified as a string - but binary data.

**Comment by InfoHunter** [comment]
> Assuming this makes it in, can this be backported into 1.1.1 or can it be written such that it can be backported into 1.1.1?

Depends on if we consider this as a *bug* or an *improvement/new feature*

**Comment by mattcaswell** [comment]
IMO this is a new feature and is not eligible for backport.

**Comment by InfoHunter** [comment]
Force-updated to support binary input by introducing a new option `sm2-bin-id`, which accepts hex string as the input. Examples of usage of the `verify` application for verifying an SM2 certificate will be either:

`openssl verify -sm2-id 1234567812345678 -CAfile SM2-CA.crt SM2.crt`

or

`openssl verify -sm2-bin-id 31323334353637383132333435363738 -CAfile SM2-CA.crt SM2.crt`

Note: document and test case are not updated yet, will finish that part if this idea is accepted by the team.

**Comment by InfoHunter** [comment]
`make update` marked several OSSL_trace functions as `NOEXIST` - is this related to my local configuration parameters of building OpenSSL?

**Comment by mspncp** [comment]
Did you reconfigure after rebasing and before running  ‚make update’? If not, try it.

**Comment by mspncp** [comment]
> ... support binary input by introducing a new option sm2-bin-id, which accepts hex string as the input. 

IMHO the name „sm2-bin-id“ is a bit misleading, because it accepts _hexadezimal_, not _binary_ input. I‘d prefer sm2-hex-id.

**Comment by InfoHunter** [comment]
> Did you reconfigure after rebasing and before running ‚make update’? If not, try it.

Yes, that works!

**Comment by InfoHunter** [comment]
Change to `sm2-hex-id`

**Comment by mspncp** [comment]
> > Did you reconfigure after rebasing and before running ‚make update’? If not, try it.
> 
> Yes, that works!

I learned that the hard way myself (with Richard's help) when rebasing #7893: whenever new files are added or existing files moved, the Makefile targets become outdated and 'make update' produces incorrect results. That's why one needs to reconfigure in order to regenerate the Makefile. I for myself adopted the habit to always reconfigure after rebasing or checking out a new branch.

File: crypto/x509/x_all.c [hunk 1]
```diff
@@ -19,10 +19,122 @@
 #include <openssl/dsa.h>
 #include <openssl/x509v3.h>
 
+#ifndef OPENSSL_NO_SM2
+
+# include "internal/asn1_int.h"
+# include "internal/evp_int.h"
+
+static int x509_verify_sm2(X509 *x, EVP_PKEY *pkey, int mdnid, int pknid);
```
Comment (t-j-h) [inline crypto/x509/x_all.c]: Why this line? The actual function is just below it. Delete.

File: crypto/x509/x_all.c [hunk 2]
```diff
@@ -19,10 +19,122 @@
...
+#ifndef OPENSSL_NO_SM2
+    int mdnid, pknid;
+#endif
+
     if (X509_ALGOR_cmp(&a->sig_alg, &a->cert_info.signature))
         return 0;
+
+#ifndef OPENSSL_NO_SM2
+    /* Convert signature OID into digest and public key OIDs */
+    if (!OBJ_find_sigid_algs(OBJ_obj2nid(a->sig_alg.algorithm),
```
Comment (t-j-h) [inline crypto/x509/x_all.c]: Do we know with certainty that all existing verify uses will have this function work correctly?
If not then its failure should not cause X509_verify itself to fail.

**Comment by InfoHunter** [comment]
> whenever new files are added or existing files moved, the Makefile targets become outdated and 'make update' produces incorrect results.

That is right. My problem is I build the code in a standalone "building" directory outside the source code tree (the .git directory) and I reconfigured in the building directory but forgot to reconfigure the source directory - at where I need to run the `make update`...

File: crypto/x509/x_all.c (hunk 1 above)
Comment (InfoHunter) [reply crypto/x509/x_all.c]: Okay

File: crypto/x509/x_all.c (hunk 2 above)
Comment (InfoHunter) [reply crypto/x509/x_all.c]: This function is called in the following `ASN1_item_verify`, so it is just moved ahead in `X509_verify`

Comment (t-j-h) [reply crypto/x509/x_all.c]: Thanks. That answers my question.

File: crypto/x509/x_all.c (hunk 1 above)
Comment (InfoHunter) [reply crypto/x509/x_all.c]: Removed in the latest commit

**Comment by InfoHunter** [comment]
force-pushed to resolve conflicts

**Comment by InfoHunter** [comment]
Travis-CI failure is not relevant. Pushed to master as 8267becb8b

File: crypto/x509/x_x509.c [hunk 3]
```diff
@@ -244,3 +244,15 @@ int X509_get_signature_nid(const X509 *x)
 {
     return OBJ_obj2nid(x->sig_alg.algorithm);
 }
+
+#ifndef OPENSSL_NO_SM2
+void X509_set_sm2_id(X509 *x, ASN1_OCTET_STRING *sm2_id)
+{
+    x->sm2_id = *sm2_id;
```
Comment (mlindner) [inline crypto/x509/x_x509.c]: This is doing a shallow copy of the `ASN1_OCTET_STRING`. If `sm2_id` is allocated on the heap and then freed you'll be left with `x` storing a `ASN1_OCTET_STRING` with it's `data` pointer deallocated which will result in memory corruption.

**Pull request closed without merging**